nowcasting_api/database_fast.py

```diff
@@ -90,3 +101,197 @@ def get_forecast_values_all_compact(
+    )
+    query = filter_gsp_id(query=query, gsp_ids=gsp_ids)
+
+    # order by target time and created utc desc
+    query = query.order_by(
+        ForecastValueLatestSQL.gsp_id,
+        ForecastValueLatestSQL.target_time,
+        ForecastValueLatestSQL.created_utc.desc(),
+    )
+
+    forecast_values = query.all()
+
+    # 3. get forecasts
+    query = session.query(
+        ForecastSQL.forecast_creation_time,
+        ForecastSQL.initialization_datetime_utc,
+        LocationSQL.gsp_id,
+        LocationSQL.gsp_name,
+        LocationSQL.gsp_group,
+        LocationSQL.region_name,
+        LocationSQL.installed_capacity_mw,
+        LocationSQL.label,
+        MLModelSQL.name,
+        MLModelSQL.version,
+        InputDataLastUpdatedSQL.gsp,
+        InputDataLastUpdatedSQL.nwp,
+        InputDataLastUpdatedSQL.pv,
+        InputDataLastUpdatedSQL.satellite,
+    )
+
+    # only get historic forecasts
+    query = query.filter(ForecastSQL.historic == True)  # noqa E712
+
+    # join with location tables
+    query = query.join(LocationSQL)
+    query = query.join(MLModelSQL)
+    query = query.join(InputDataLastUpdatedSQL)
+
+    # filters
+    query = filter_gsp_id(gsp_ids=gsp_ids, query=query, model=LocationSQL)
+    query = query.filter(ForecastSQL.model_id.in_(model_ids))
+
+    # order by gsp_id
+    query = query.order_by(LocationSQL.gsp_id)
+
+    forecasts = query.all()
+
+    # 4. Convert to Forecast objects
+    forecast_objects = {}
+    for forecast in forecasts:
+        forecast_creation_time = forecast[0]
+        initialization_datetime_utc = forecast[1]
+        gsp_id = forecast[2]
+
+        # creation Location object
+        location = Location(
+            gsp_id=gsp_id,
+            gsp_name=forecast[3],
+            gsp_group=forecast[4],
+            region_name=forecast[5],
+            installed_capacity_mw=forecast[6],
+            label=forecast[7],
+        )
+
+        # create model object
+        model = MLModel(name=forecast[8], version=forecast[9])
+
+        # create input last update object
+        input_data_last_updated = InputDataLastUpdated(
+            gsp=forecast[10],
+            nwp=forecast[11],
+            pv=forecast[12],
+            satellite=forecast[13],
+        )
+        forecast_py = Forecast(
+            historic=True,
+            forecast_creation_time=forecast_creation_time,
+            initialization_datetime_utc=initialization_datetime_utc,
+            forecast_values=[],
+            location=location,
+            model=model,
+            input_data_last_updated=input_data_last_updated,
+        )
+        forecast_objects[gsp_id] = forecast_py
+
+    # 5. add forecast values to forecast objects
+    for forecast_value in forecast_values:
+        datetime_utc = forecast_value[0]
+        power_mw = forecast_value[1]
+        gsp_id = forecast_value[2]
+
+        power_mw = round(power_mw, 2)
```

Comment:
could be done in database query

Reply:
100%

Reply:
I think this is the last thing on this PR then good to merge

Reply:
done, thanks